Extracting external sites

bmenn opened the pull request:
We need functionality to identify other potential sites to scrape.

bmenn committed:
Add TravisCI configuration and related badges

Adding TravisCI to ensure tests are run and tests are passing before
PR's are merged.

bmenn committed:
Fix TravisCI test runner command

bmenn committed:
Fix typos in TravisCI configuration and README badges

bmenn committed:
WIP Adding Luigi task to extract external sites

Need to add the ability to extract external sites so that we can filter
these for potential new sites to scrape.

bmenn commented:
If someone get to this before I get back to this, we could use SQLAlchemy data model objects to simplify data access (but this does not mean less code).

http://docs.sqlalchemy.org/en/latest/orm/extensions/automap.html?highlight=automap#module-sqlalchemy.ext.automap

bmenn committed:
Add URL authority (website) extraction functionality

bmenn committed:
Finish Luigi task to extract external sites from a page

By extracting external sites from a page, we can identify other
additional websites of interest to scrape. This also provides an example
on how to unit test Luigi tasks.

bmenn commented:
@dlrobertson @egrossman 

Not 100% done here, but almost. Need to add a simple integration test and I'll be good. There's also an example of unit testing a Luigi task as well.

bmenn committed:
WIP Created basic integration test for BaseSites task

Created test to confirm queries are functioning as expected. Had to
reorganize the test suite to easily separate unit tests from integration
tests, to allow control which set is run by filtering on directories.

bmenn committed:
Finish integration tests, Vagrantfile WIP

bmenn committed:
Fix failing test import

dlrobertson commented on Vagrantfile:
```diff
@@ -0,0 +1,74 @@
+# -*- mode: ruby -*-
+# vi: set ft=ruby :
```
:+1:

dlrobertson commented on Vagrantfile:
```diff
@@ -0,0 +1,74 @@
+    apt-get update
+    apt-get install -y postgresql postgresql-contrib
+    echo "listen_addresses = '*'" | sudo tee -a /etc/postgresql/9.5/main/postgresql.conf
+    echo "host all postgres 0.0.0.0/0 trust" | sudo tee /etc/postgresql/9.5/main/pg_hba.conf
```
Do we really need it this open? I'd think `localhost` would be good enough for both listen and no password auth.

bmenn replied:
We need this because localhost isn't the dev's localhost when using Vagrant.

dlrobertson commented on htetl/tasks/sites.py:
```diff
@@ -0,0 +1,52 @@
+                'PageId': row.id,
+                'ExternalSiteId': base_sites.iloc[external_site]['Id']
+                }
+                for external_site in site_list
```
:boom:

dlrobertson commented on tests/integration/__init__.py:
```diff
@@ -0,0 +1,59 @@
+if sys.version_info[0] < 3:
+    from StringIO import StringIO
+else:
+    from io import StringIO
```
Is this the more pythonic way? I thought the standard way of doing this was

```python
try:
    import x
except ImportError:
    import y
````

bmenn replied:
Yes, that is more Pythonic. Stole the line from SO without thinking about that.

dlrobertson commented on tests/integration/__init__.py:
```diff
@@ -0,0 +1,59 @@
+    test_config.set('QueryPostgres', 'user', POSTGRES_USER)
+    test_config.set('QueryPostgres', 'password', POSTGRES_PASSWORD)
+
+    # meta = sqlalchemy.MetaData(engine)
```
Is this going to be used. If not I'm a fan of just removing it

bmenn replied:
Should be gone now.

dlrobertson commented on tests/integration/__init__.py:
```diff
@@ -0,0 +1,59 @@
+    ).to_sql('BaseSites', engine)
+#
+#
+# def teardown_package():
```
Same

dlrobertson commented on tests/unit/tasks/mocks.py:
```diff
@@ -0,0 +1,7 @@
+from htetl.tasks.loadpages import RawPageData
+
+
+def setup_RawPageData():
```
Isn't a `pass` needed here?

bmenn replied:
Honestly, I don't know why this file is here.

bmenn committed:
Fix old bug in QueryPostgres for CSV writing with working Vagrantfile

Updated QueryPostgres to write the entire CSV file instead of just
appending to the file.

Vagrantfile now seems to be functioning well for tests.

bmenn committed:
Undo intentionally failing test

bmenn committed:
Address PR comments and cleaned up some

egrossman commented on tests/unit/tasks/test_sites.py:
```diff
@@ -0,0 +1,33 @@
+    test_config.set('QueryPostgres', 'database', 'fake_db')
+    test_config.set('QueryPostgres', 'user', 'fake_user')
+    test_config.set('QueryPostgres', 'password', 'fake_password')
+
```
Is setup_module used anywhere or is this ran automatically by luigi?

bmenn replied:
`setup_module` is used by nose. A little more details here if you are interested http://nose.readthedocs.io/en/latest/writing_tests.html.

egrossman reviewed (approved):
It took me way too long to figure out how mock testing works, but finally got through the code. I noticed that extacts.sites isn't a luigi task, but just methods to be called in tasks. Is that how you want us to refactor our code too?
